bigquery/src/test/java/org/apache/iceberg/gcp/bigquery/BigQueryTableOperationsTest.java

```diff
@@ -0,0 +1,274 @@
+      mock(BigQueryMetastoreClient.class);
+
+  private BigQueryMetastoreCatalog bigQueryMetastoreCatalog;
+  private BigQueryTableOperations tableOps;
+
+  @BeforeEach
+  public void before() {
+    this.bigQueryMetastoreCatalog = new BigQueryMetastoreCatalog();
+    this.bigQueryMetastoreCatalog.setConf(new Configuration());
+    String warehouseLocation = tempFolder.toPath().resolve("hive-warehouse").toString();
+
+    bigQueryMetastoreCatalog.initialize(
+        "CATALOG_ID",
+        ImmutableMap.of(
+            PROJECT_ID,
+            GCP_PROJECT,
+            CatalogProperties.WAREHOUSE_LOCATION,
+            warehouseLocation,
+            CatalogProperties.FILE_IO_IMPL,
+            "org.apache.iceberg.hadoop.HadoopFileIO"),
+        GCP_PROJECT,
+        GCP_REGION,
+        bigQueryMetaStoreClient);
+    this.tableOps = (BigQueryTableOperations) bigQueryMetastoreCatalog.newTableOps(SPARK_TABLE_ID);
+  }
+
+  @Test
+  public void testFetchLatestMetadataFromBigQuery() throws Exception {
+    Table createdTable = createTestTable();
+    reset(bigQueryMetaStoreClient);
+    when(bigQueryMetaStoreClient.load(TABLE_REFERENCE)).thenReturn(createdTable);
+
+    tableOps.refresh();
+    assertThat(
+            createdTable
+                .getExternalCatalogTableOptions()
+                .getParameters()
+                .getOrDefault(METADATA_LOCATION_PROP, ""))
+        .isEqualTo(tableOps.currentMetadataLocation());
+
+    reset(bigQueryMetaStoreClient);
+    when(bigQueryMetaStoreClient.load(TABLE_REFERENCE))
+        .thenThrow(new NoSuchTableException("error message getTable"));
+    // Refresh fails when table is not found but metadata already presents.
+    assertThatThrownBy(() -> tableOps.refresh())
+        .isInstanceOf(NoSuchTableException.class)
+        .hasMessageContaining("error message getTable");
+  }
+
+  @Test
+  public void testFailForNonIcebergTable() {
+    when(bigQueryMetaStoreClient.load(TABLE_REFERENCE))
+        .thenReturn(new Table().setTableReference(TABLE_REFERENCE));
+
+    assertThatThrownBy(() -> tableOps.refresh())
+        .isInstanceOf(ValidationException.class)
+        .hasMessageContaining("metadata location not found");
+  }
+
+  @Test
+  public void testNoOpWhenMetadataAndTableNotFound() {
+    when(bigQueryMetaStoreClient.load(TABLE_REFERENCE))
+        .thenThrow(new NoSuchTableException("error message getTable"));
+    // Table not found won't cause errors when the metadata is null.
+    assertThat(tableOps.currentMetadataLocation()).isNull();
+    tableOps.refresh();
+  }
+
+  @Test
+  public void testTableNameAsExpected() {
+    assertThat(tableOps.tableName()).isEqualTo("db.tbl");
+  }
+
+  @Test
+  public void testUseEtagForUpdateTable() throws Exception {
+    Table tableWithEtag = createTestTable().setEtag("etag");
+    reset(bigQueryMetaStoreClient);
+    when(bigQueryMetaStoreClient.load(TABLE_REFERENCE)).thenReturn(tableWithEtag, tableWithEtag);
```

Comment:
```suggestion
    when(bigQueryMetaStoreClient.load(TABLE_REFERENCE)).thenReturn(tableWithEtag);
```

I don't think the second one is actually needed